README.md
# ca.taffer.mm-rolly

Dice rolling plugin for Mattermost.

![Dice icon](if_die_1055072.png)

Inspired by
[moussetc/mattermost-plugin-dice-roller](https://github.com/moussetc/mattermost-plugin-dice-roller)
and [DiceBot](https://dice-b.appspot.com/) for Slack.

Note that I don't actually know Go, so this could be rough...

## Rolling Dice

Support "any" [reasonable](https://en.wikipedia.org/wiki/Dice_notation) dice
rolling request:

* *x*d*y* or *x*D*y* to roll a *y* sided die *x* times
* modifiers: *x*d*y*+*z* (supported modifiers: +, -, x or *, /)
* exploding dice (for every max value, roll and add): *x*d*y*!
* *x*d% - same as *x*d100
* *x*dF - roll
  [FUDGE](https://en.wikipedia.org/wiki/Fudge_%28role-playing_game_system%29)
  dice
* *x*d*y*<*z* - discards the lowest *z* rolls (so 4d6<1 would return a value
  between 3 and 18)
* *x*d*y*>*z* - keeps the best *z* rolls (so 4d6>1 would return a value
  between 1 and 6)

If *x* isn't specified, it defaults to 1. If *y* is less than 2, it defaults
to 2. If you specify a modifier, you must also specify a *z* value.

Nerd combos:

* dnd - same as 3d6 six times (standard D&D or Pathfinder)
* dnd+ - same as 4d6<1 six times (common house rule for D&D or Pathfinder)
* open - roll d%, if it's >= 95, roll again and add, repeating if necessary

Number of dice per roll will be limited to 100 so malicious users can't flood
the channel with dice output.

Number of rolls per request (`/roll 1d6 2d6 ... n`) will be limited to 10 so
malicious users can't flood the channel with dice output.

## Credits

* [Mattermost's plugin sample](https://github.com/mattermost/mattermost-plugin-sample)
* Icon from [Nick Roach](http://www.elegantthemes.com/)'s GPL'd
  [Circle Icons](https://www.iconfinder.com/iconsets/circle-icons-1) set.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [pre-1.0] - 2018-08-16

### Changed

* Updated
  [Mattermost's plugin sample](https://github.com/mattermost/mattermost-plugin-sample)
  to customize it for Rolly. (#7)


server/plugin_test.go
package main

import (
	"io/ioutil"
	"math/rand"
	"net/http/httptest"
	"regexp"
	"strings"
	"testing"

	"github.com/mattermost/mattermost-server/model"
	"github.com/mattermost/mattermost-server/plugin"
	"github.com/mattermost/mattermost-server/plugin/plugintest"
	"github.com/mattermost/mattermost-server/plugin/plugintest/mock"
	"github.com/stretchr/testify/assert"
)

// -----------------------------------------------------------------------------
// Utilities
// -----------------------------------------------------------------------------

func runTestPluginCommand(t *testing.T, cmd string) (*model.CommandResponse, *model.AppError) {
	p := initTestPlugin(t)
	assert.Nil(t, p.OnActivate())

	var command *model.CommandArgs
	command = &model.CommandArgs{
		Command: cmd,
	}

	return p.ExecuteCommand(&plugin.Context{}, command)
}

func initTestPlugin(t *testing.T) *RollyPlugin {
	api := &plugintest.API{}
	api.On("RegisterCommand", mock.Anything).Return(nil)
	api.On("UnregisterCommand", mock.Anything, mock.Anything).Return(nil)
	api.On("GetUser", mock.Anything).Return(&model.User{
		Id:        "userid",
		Nickname:  "User",
		Username:  "hunter2",
		FirstName: "User",
		LastName:  "McUserface",
	}, (*model.AppError)(nil))

	p := RollyPlugin{}
	p.SetAPI(api)

	return &p
}

// -----------------------------------------------------------------------------
// Tests - Mattermost Plugin Interface
// -----------------------------------------------------------------------------

// TestPluginLifecycle - Make sure OnActivate()/OnDeactivate() don't crash.
func TestPluginLifecycle(t *testing.T) {
	p := initTestPlugin(t)

	assert.Nil(t, p.OnActivate())
	assert.True(t, p.active)

	// TODO : Try ExecuteCommand while deactivated?

	assert.Nil(t, p.OnDeactivate())
	assert.False(t, p.active)
}

// TestServeHTTP - Test the HTTP response.
func TestServeHTTP(t *testing.T) {
	assert := assert.New(t)
	plugin := RollyPlugin{}
	w := httptest.NewRecorder()
	r := httptest.NewRequest("GET", "/", nil)

	plugin.ServeHTTP(nil, w, r)

	result := w.Result()
	assert.NotNil(result)
	bodyBytes, err := ioutil.ReadAll(result.Body)
	assert.Nil(err)
	bodyString := string(bodyBytes)

	assert.Equal("Please log in.\n", bodyString)
}

// -----------------------------------------------------------------------------
// Tests - Plugin utilities.
// -----------------------------------------------------------------------------

// TestGetHelp - Help command returns help text.
func TestGetHelp(t *testing.T) {
	p := initTestPlugin(t)
	resp, err := p.GetHelp()

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// TODO: Can these strings be moved out into something useful for
	// i18n?
	assert.True(t, strings.HasPrefix(resp.Text, `Support "any" [reasonable](https://en.wikipedia.org/wiki/Dice_notation) dice rolling request:`))
}

// TestInit - Make sure Init() compiles the regexes.
func TestInit(t *testing.T) {
	p := initTestPlugin(t)

	assert.Nil(t, p.simplePattern)
	assert.Nil(t, p.comboPattern)
	assert.Nil(t, p.rollPattern)

	p.Init()

	assert.NotNil(t, p.simplePattern)
	assert.NotNil(t, p.comboPattern)
	assert.NotNil(t, p.rollPattern)
}

// TestGetCommand - How's this going to fail, really?
func TestGetCommand(t *testing.T) {
	p := initTestPlugin(t)

	cmd := p.GetCommand()

	assert.EqualValues(t, cmd.Trigger, "roll")
	assert.EqualValues(t, cmd.Description, "Roll one or more dice. With combos!")
}

// TestHandleRoll - Make sure different sorts of rolls do things.
func TestHandleRoll(t *testing.T) {
	p := initTestPlugin(t)
	p.Init()

	// simplePattern matches
	rand.Seed(0) // Make these deterministic.
	response := p.HandleRoll("6", "")
	assert.EqualValues(t, response, `"1d6" = **1**`)

	response = p.HandleRoll("%", "")
	assert.EqualValues(t, response, `"1d%" = **15**`)

	response = p.HandleRoll("F", "")
	assert.EqualValues(t, response, `"1dF" = **0**`)

	// comboPattern matches
	rand.Seed(0) // Make these deterministic.
	response = p.HandleRoll("dnd", "")
	assert.EqualValues(t, response, "D&D standard:\n* 3d6 [1 1 2] = **4**\n* 3d6 [5 5 6] = **16**\n* 3d6 [1 2 6] = **9**\n* 3d6 [1 1 6] = **8**\n* 3d6 [1 1 6] = **8**\n* 3d6 [1 3 6] = **10**")

	response = p.HandleRoll("dnd+", "")
	assert.EqualValues(t, response, "D&D variant:\n* 4d6<1 [1 1 5 6] = **12**\n* 4d6<1 [1 1 3 6] = **10**\n* 4d6<1 [1 2 5 5] = **12**\n* 4d6<1 [3 4 5 6] = **15**\n* 4d6<1 [2 3 3 5] = **11**\n* 4d6<1 [1 1 2 6] = **9**")

	response = p.HandleRoll("open", "")
	assert.EqualValues(t, response, "Rolemaster open-ended: 1d% [77] = **77**")

	// rollPattern matches
	rand.Seed(0) // Make these deterministic.
	response = p.HandleRoll("d3", "")
	assert.EqualValues(t, response, `"d3" = **1**`)

	response = p.HandleRoll("2d3", "")
	assert.EqualValues(t, response, `"2d3" [1 2] = **3**`)

	response = p.HandleRoll("1000d6", "")
	assert.EqualValues(t, response, "1000 is too many, rolling 100.\n\"1000d6\" [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4 4 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 5 6 6 6 6 6 6 6 6 6 6 6 6 6 6 6] = **320**")

	response = p.HandleRoll("6!", "")
	assert.EqualValues(t, response, "\"6!\" [6 1] = **7**")

	response = p.HandleRoll("monkey", "")
	assert.EqualValues(t, response, "I have no idea what to do with this: monkey")

	response = p.HandleRoll("1", "")
	assert.EqualValues(t, response, "Your one-sided die rolls off into the shadows.")

	response = p.HandleRoll("1d1", "")
	assert.EqualValues(t, response, "Your one-sided die rolls off into the shadows.")
}

// TestRollDice - Make sure different combinations return correct values.
func TestRollDice(t *testing.T) {
	p := initTestPlugin(t)
	p.Init()

	// Different number of sides.
	rand.Seed(0) // Make these deterministic.
	rolls, total := p.RollDice(1, "%", "", 0)
	assert.EqualValues(t, rolls[0], 75)
	assert.EqualValues(t, total, 75)

	rolls, total = p.RollDice(1, "F", "", 0)
	assert.EqualValues(t, rolls[0], -1)
	assert.EqualValues(t, total, -1)

	rolls, total = p.RollDice(1, "1", "", 0)
	assert.EqualValues(t, rolls[0], 2)
	assert.EqualValues(t, total, 2)

	// Different modifiers.
	rand.Seed(0) // Make these deterministic.
	rolls, total = p.RollDice(1, "6", "+", 1)
	assert.EqualValues(t, rolls[0], 1)
	assert.EqualValues(t, total, 2)

	rolls, total = p.RollDice(1, "6", "-", 6)
	assert.EqualValues(t, rolls[0], 1)
	assert.EqualValues(t, total, 1)

	rolls, total = p.RollDice(1, "6", "/", 2)
	assert.EqualValues(t, rolls[0], 2)
	assert.EqualValues(t, total, 1)

	rolls, total = p.RollDice(1, "6", "x", 2)
	assert.EqualValues(t, rolls[0], 5)
	assert.EqualValues(t, total, 10)

	rand.Seed(0) // Make these deterministic.
	rolls, total = p.RollDice(2, "6", "<", 1)
	assert.EqualValues(t, rolls[0], 1)
	assert.EqualValues(t, rolls[1], 1)
	assert.EqualValues(t, total, 1)

	rolls, total = p.RollDice(2, "6", ">", 1)
	assert.EqualValues(t, rolls[0], 2)
	assert.EqualValues(t, rolls[1], 5)
	assert.EqualValues(t, total, 5)

	rolls, total = p.RollDice(2, "6", ">", 3)
	assert.EqualValues(t, rolls[0], 5)
	assert.EqualValues(t, rolls[1], 6)
	assert.EqualValues(t, total, 11)

	rolls, total = p.RollDice(2, "6", ">", 0)
	assert.EqualValues(t, rolls[0], 2)
	assert.EqualValues(t, rolls[1], 6)
	assert.EqualValues(t, total, 0)

	// Need to roll these several times before it actually explodes.
	rand.Seed(0) // Make these deterministic.
	rolls, total = p.RollDice(1, "F", "!", 0)
	rolls, total = p.RollDice(1, "F", "!", 0)
	rolls, total = p.RollDice(1, "F", "!", 0)
	rolls, total = p.RollDice(1, "F", "!", 0)
	rolls, total = p.RollDice(1, "F", "!", 0)
	assert.EqualValues(t, rolls[0], 1)
	assert.EqualValues(t, rolls[1], 0)
	assert.EqualValues(t, total, 1)
}

// TestSeedRng - Seed the default RNG.
func TestSeedRng(t *testing.T) {
	p := initTestPlugin(t)
	p.SeedRng()

	assert.NotNil(t, p) // Not actually testing anything useful.
}

// TestGetRandom - Make sure GetRandom() is returning values.
func TestGetRandom(t *testing.T) {
	p := initTestPlugin(t)
	rand.Seed(0) // Make these deterministic.

	val := p.GetRandom(1)
	assert.EqualValues(t, val, 1)

	val = p.GetRandom(6)
	assert.True(t, val >= 1 && val <= 6)
	assert.EqualValues(t, val, 1)
}

// TestGetName - Make sure GetName() does something suitable.
func TestGetName(t *testing.T) {
	p := initTestPlugin(t)

	// TODO: How to mock this better so we can test users with missing fields?
	name, err := p.GetName("userid")
	assert.Nil(t, err)
	assert.EqualValues(t, name, "User")
}

// -----------------------------------------------------------------------------
// Tests - Plugin triggers.
// -----------------------------------------------------------------------------

// TestWrongTrigger - Wrong trigger shouldn't do anything.
func TestWrongTrigger(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/ihavenoideawhatimdoing")

	assert.Nil(t, resp)
	assert.Nil(t, err)
}

// TestRollHelp - Help command returns help text.
func TestRollHelp(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll help")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// TODO: Can these strings be moved out into something useful for
	// i18n?
	assert.True(t, strings.HasPrefix(resp.Text, `Support "any" [reasonable](https://en.wikipedia.org/wiki/Dice_notation) dice rolling request:`))
}

// TestRollEmpty - Empty /roll returns a snide comment.
func TestRollEmpty(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// Negative tests.
	assert.False(t, strings.Contains(resp.Text, "rolls requested; I'm only doing 10."))

	// Positive tests.
	assert.True(t, strings.Contains(resp.Text, "throws the dice…"))
	assert.True(t, strings.Contains(resp.Text, "That accomplished nothing."))
}

// TestRollTooMany - /roll with too many requests includes a warning.
func TestRollTooMany(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll 1 2 3 4 5 6 7 8 9 10 11")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// Negative tests.
	assert.False(t, strings.Contains(resp.Text, "That accomplished nothing."))

	// Positive tests.
	assert.True(t, strings.Contains(resp.Text, "throws the dice…"))
	assert.True(t, strings.Contains(resp.Text, "rolls requested; I'm only doing 10."))
}

// TestSimpleRoll - /roll with a number.
func TestSimpleRoll(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll 6")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// Negative tests.
	assert.False(t, strings.Contains(resp.Text, "That accomplished nothing."))
	assert.False(t, strings.Contains(resp.Text, "rolls requested; I'm only doing 10."))

	// Positive tests.
	assert.True(t, strings.Contains(resp.Text, "throws the dice…"))
	assert.True(t, strings.Contains(resp.Attachments[0].Text, "1d6"))
}

// TestNormalRoll - /roll with an xdy value.
func TestNormalRoll(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll 3d6")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// Negative tests.
	assert.False(t, strings.Contains(resp.Text, "That accomplished nothing."))
	assert.False(t, strings.Contains(resp.Text, "rolls requested; I'm only doing 10."))

	// Positive tests.
	assert.True(t, strings.Contains(resp.Text, "throws the dice…"))
	assert.True(t, strings.Contains(resp.Attachments[0].Text, "3d6"))
}

// TestComboRoll - /roll with a combo.
func TestComboRoll(t *testing.T) {
	resp, err := runTestPluginCommand(t, "/roll dnd")

	assert.NotNil(t, resp)
	assert.Nil(t, err)

	// Negative tests.
	assert.False(t, strings.Contains(resp.Text, "That accomplished nothing."))
	assert.False(t, strings.Contains(resp.Text, "rolls requested; I'm only doing 10."))

	// Positive tests.
	assert.True(t, strings.Contains(resp.Text, "throws the dice…"))
	assert.True(t, strings.Contains(resp.Attachments[0].Text, "D&D standard:"))
}

// -----------------------------------------------------------------------------
// Tests - Other functions.
// -----------------------------------------------------------------------------

// TestSum - Make sure sum() sums the contents.
func TestSum(t *testing.T) {
	assert.EqualValues(t, sum([]int{}), 0)
	assert.EqualValues(t, sum([]int{1}), 1)
	assert.EqualValues(t, sum([]int{1, -1}), 0)
	assert.EqualValues(t, sum([]int{1, 2, 3}), 6)
}

// TestMin - Make sure min returns the minimum.
func TestMin(t *testing.T) {
	assert.EqualValues(t, min(1, 2), 1)
	assert.EqualValues(t, min(2, 1), 1)
}

// TestFindNamedSubstrings - Make sure regexes can be turned into dicts.
//
// This assumes you've already checked to see if there's a match inside.
func TestFindNamedSubstrings(t *testing.T) {
	// Test with the simpleRegex.
	pattern := regexp.MustCompile(simpleRegex)
	matches := FindNamedSubstrings(pattern, "6")
	assert.EqualValues(t, matches["num_sides"], "6")

	// Test with the comboRegex.
	pattern = regexp.MustCompile(comboRegex)
	matches = FindNamedSubstrings(pattern, "dnd")
	assert.EqualValues(t, matches["combo_name"], "dnd")

	// Test with the rollRegex.
	pattern = regexp.MustCompile(rollRegex)
	matches = FindNamedSubstrings(pattern, "1d2+3")
	assert.EqualValues(t, matches["num_dice"], "1")
	assert.EqualValues(t, matches["num_sides"], "2")
	assert.EqualValues(t, matches["modifier"], "+")
	assert.EqualValues(t, matches["modifier_value"], "3")
}


server/configuration.go
package main

import (
	"github.com/pkg/errors"
)

// -----------------------------------------------------------------------------
// Constants
// -----------------------------------------------------------------------------

const defaultTrigger string = "rolly"

// configuration captures the plugin's external configuration as exposed in the Mattermost server
// configuration, as well as values computed from the configuration. Any public fields will be
// deserialized from the Mattermost server configuration in OnConfigurationChange.
//
// As plugins are inherently concurrent (hooks being called asynchronously), and the plugin
// configuration can change at any time, access to the configuration must be synchronized. The
// strategy used in this plugin is to guard a pointer to the configuration, and clone the entire
// struct whenever it changes. You may replace this with whatever strategy you choose.
//
// If you add non-reference types to your configuration struct, be sure to rewrite Clone as a deep
// copy appropriate for your types.
type configuration struct {
	trigger string
}

// Clone shallow copies the configuration. Your implementation may require a deep copy if
// your configuration has reference types.
func (c *configuration) Clone() *configuration {
	var clone = *c
	return &clone
}

// getConfiguration retrieves the active configuration under lock, making it safe to use
// concurrently. The active configuration may change underneath the client of this method, but
// the struct returned by this API call is considered immutable.
func (p *RollyPlugin) getConfiguration() *configuration {
	p.configurationLock.RLock()
	defer p.configurationLock.RUnlock()

	if p.configuration == nil {
		return &configuration{}
	}

	return p.configuration
}

// setConfiguration replaces the active configuration under lock.
//
// Do not call setConfiguration while holding the configurationLock, as sync.Mutex is not
// reentrant. In particular, avoid using the plugin API entirely, as this may in turn trigger a
// hook back into the plugin. If that hook attempts to acquire this lock, a deadlock may occur.
//
// This method panics if setConfiguration is called with the existing configuration. This almost
// certainly means that the configuration was modified without being cloned and may result in
// an unsafe access.
func (p *RollyPlugin) setConfiguration(configuration *configuration) {
	p.configurationLock.Lock()
	defer p.configurationLock.Unlock()

	if configuration != nil && p.configuration == configuration {
		panic("setConfiguration called with the existing configuration")
	}

	// I'm not sure if we need to do this in both setConfiguration() and
	// OnConfigurationChange() to safely default it when there are no config
	// values in the server's config file.
	if configuration.trigger == "" {
		configuration.trigger = defaultTrigger
	}

	p.configuration = configuration
}

// OnConfigurationChange is invoked when configuration changes may have been made.
func (p *RollyPlugin) OnConfigurationChange() error {
	var configuration = new(configuration)

	// Load the public configuration fields from the Mattermost server configuration.
	if err := p.API.LoadPluginConfiguration(configuration); err != nil {
		return errors.Wrap(err, "failed to load plugin configuration")
	}

	if configuration.trigger == "" {
		configuration.trigger = defaultTrigger
	}

	p.setConfiguration(configuration)

	return nil
}


server/plugin.go
package main

import (
	"fmt"
	"math/rand"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/gorilla/mux"
	"github.com/mattermost/mattermost-server/model"
	"github.com/mattermost/mattermost-server/plugin"
)

// RollyPlugin instance.
type RollyPlugin struct {
	plugin.MattermostPlugin

	// URL router/dispatcher.
	router *mux.Router

	// Is this active?
	active bool

	// Dice rolling patterns.
	simplePattern *regexp.Regexp
	comboPattern  *regexp.Regexp
	rollPattern   *regexp.Regexp
}

// -----------------------------------------------------------------------------
// Constants
// -----------------------------------------------------------------------------

const (
	pluginPath string = "plugins/ca.taffer.mm-rolly" // Maybe?
	iconFile   string = "if_die_1055072.png"
	iconPath   string = pluginPath + "/" + iconFile
	iconURI    string = "/" + iconPath

	trigger    string = "roll"
	pluginName string = "Rolly"

	simpleRegex string = `^(?P<num_sides>[0-9\%F]+)$`
	comboRegex  string = `(?i)^((?P<combo_name>(d[n&]d\+?|open)))$`
	rollRegex   string = `(?i)^((?P<num_dice>[0-9]+)?d)?(?P<num_sides>[0-9\%F]+)((?P<modifier>[+-/<>x*!])(?P<modifier_value>[0-9]*))?$`
)

// -----------------------------------------------------------------------------
// Mattermost events.
// -----------------------------------------------------------------------------

// OnActivate - Register the plugin.
//
// Is this called once when the plugin is loaded? Repeatedly? Whenever you
// switch from "Deactivated" to "Activated"?
func (p *RollyPlugin) OnActivate() error {
	p.active = true

	p.Init()

	// Handle requests for the icon file.  Maybe this nil check isn't necessary
	// and OnActivate() is only called once per lifetime?
	if p.router == nil {
		p.router = mux.NewRouter()
		p.router.HandleFunc("/"+iconFile, func(w http.ResponseWriter, r *http.Request) {
			http.ServeFile(w, r, iconPath)
		})
	}

	// Register our command handler.
	err := p.API.RegisterCommand(p.GetCommand())

	return err
}

// OnDeactivate - No more work...
func (p *RollyPlugin) OnDeactivate() error {
	p.active = false

	return nil
}

// ServeHTTP - Handle HTTP requests.
func (p *RollyPlugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
	if r.Header.Get("Mattermost-User-Id") == "" {
		http.Error(w, "Please log in.", http.StatusForbidden)
		return
	}

	p.router.ServeHTTP(w, r)
}

// ExecuteCommand - Handle commands.
func (p *RollyPlugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
	if strings.HasPrefix(args.Command, "/"+trigger) == false {
		// It's not for us.
		return nil, nil
	}

	if strings.Contains(args.Command, "help") {
		return p.GetHelp()
	}

	// Get the user to we can display the right name.
	userName, userErr := p.GetName(args.UserId)
	if userErr != nil {
		return nil, userErr
	}

	responseText := fmt.Sprintf("%s throws the dice…", userName)

	rolls := strings.Fields(args.Command)[1:]
	if len(rolls) > 10 {
		rolls = rolls[0:11]
		responseText += fmt.Sprintf("\n⚠️ %d rolls requested; I'm only doing 10.", len(rolls))
	}

	attachments := []*model.SlackAttachment{}
	if len(rolls) == 0 {
		responseText += fmt.Sprintf("\n🚫 That accomplished nothing.")
	} else {
		rollText := ""
		for idx := 0; idx < len(rolls); idx++ {
			rollText += "\n🎲 "
			rollText = p.HandleRoll(rolls[idx], rollText)
		}

		attachments = []*model.SlackAttachment{
			{
				Text:     rollText,
				Fallback: "🎲",
				ThumbURL: iconURI,
			},
		}
	}

	props := map[string]interface{}{
		"from_webhook":  "true",
		"use_user_icon": "true",
	}

	return &model.CommandResponse{
		ResponseType: model.COMMAND_RESPONSE_TYPE_IN_CHANNEL,
		Attachments:  attachments,
		Username:     pluginName,
		Text:         responseText,
		Props:        props,
		IconURL:      iconURI,
	}, nil
}

// -----------------------------------------------------------------------------
// Different commands the roller knows.
// -----------------------------------------------------------------------------

// GetHelp - Produce a help message when the user asks for help.
func (p *RollyPlugin) GetHelp() (*model.CommandResponse, *model.AppError) {
	helpText := `Support "any" [reasonable](https://en.wikipedia.org/wiki/Dice_notation) dice rolling request:

* *x*d*y* or *x*D*y* to roll a *y* sided die *x* times
* modifiers: *x*d*y*+*z* (supported modifiers: +, -, x or *, /)
* exploding dice (for every max value, roll and add): *x*d*y*!
* *x*d% - same as *x*d100
* *x*dF - roll
  [FUDGE](https://en.wikipedia.org/wiki/Fudge_%28role-playing_game_system%29)
  dice
* *x*d*y*<*z* - discards the lowest *z* rolls (so 4d6<1 would return a value
  between 3 and 18)
* *x*d*y*>*z* - keeps the best *z* rolls (so 4d6>1 would return a value
  between 1 and 6)

If *x* isn't specified, it defaults to 1. If *y* is less than 2, it defaults
to 2. If you specify a modifier, you must also specify a *z* value.

Supports these nerd combos:

* dnd - same as 3d6 six times (standard D&D or Pathfinder)
* dnd+ - same as 4d6<1 six times (common house rule for D&D or Pathfinder)
* open - roll d%, if it's >= 95, roll again and add, repeating if necessary`

	props := map[string]interface{}{
		"from_webhook":  "true",
		"use_user_icon": "true",
	}

	return &model.CommandResponse{
		ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL,
		Text:         helpText,
		Props:        props,
		Username:     pluginName,
		IconURL:      iconURI,
	}, nil
}

// HandleRoll - Handle a rolling command.
//
// Returns the adjusted roll output.
func (p *RollyPlugin) HandleRoll(rollArg string, rollText string) string {
	if p.simplePattern.MatchString(rollArg) == true {
		// Simple roll (number only).
		matches := FindNamedSubstrings(p.simplePattern, rollArg)

		if matches["num_sides"] == "1" {
			rollText += "Your one-sided die rolls off into the shadows."
		} else {
			dice, total := p.RollDice(1, matches["num_sides"], "", 0)

			if len(dice) == 1 {
				rollText += fmt.Sprintf("\"1d%v\" = **%d**", rollArg, total)
			} else {
				rollText += fmt.Sprintf("%q %v = **%d**", rollArg, dice, total)
			}
		}

	} else if p.comboPattern.MatchString(rollArg) == true {
		// C-C-C-C-COMBO roll.
		matches := FindNamedSubstrings(p.comboPattern, rollArg)

		comboName := strings.ToLower(matches["combo_name"])
		switch comboName {
		case "dnd", "d&d":
			// D&D/Pathfinder: 3d6 for each stat.
			rollText += "D&D standard:"

			for idx := 0; idx < 6; idx++ {
				dice, total := p.RollDice(3, "6", "", 0)
				rollText += fmt.Sprintf("\n* 3d6 %v = **%d**", dice, total)
			}
		case "dnd+", "d&d+":
			// Common D&D/Pathfinder house rule: 4d6<1 for each stat.
			rollText += "D&D variant:"

			for idx := 0; idx < 6; idx++ {
				dice, total := p.RollDice(4, "6", "<", 1)
				rollText += fmt.Sprintf("\n* 4d6<1 %v = **%d**", dice, total)
			}
		case "open":
			// Rolemaster open-ended d%.
			dice, total := p.RollDice(1, "%", "", 0)
			allDice := dice
			for total >= 95 {
				dice, total = p.RollDice(1, "%", "", 0)
				allDice = append(allDice, dice[0])
			}

			sort.Ints(allDice)
			total = sum(allDice)
			rollText += fmt.Sprintf("Rolemaster open-ended: 1d%% %v = **%d**", allDice, total)
		default:
			// You can't actually reach this with the current regex.
			rollText += fmt.Sprintf("Combo **%v** isn't implemented yet, sorry.", rollArg)
		}

	} else if p.rollPattern.MatchString(rollArg) == true {
		// Typical roll (number of dice, sides, optional modifiers).
		matches := FindNamedSubstrings(p.rollPattern, rollArg)

		numDice, err := strconv.Atoi(matches["num_dice"])
		if err != nil {
			// This is optional, so it might be empty.
			numDice = 1
		}
		if numDice > 100 {
			rollText += fmt.Sprintf("%v is too many, rolling 100.\n", numDice)
			numDice = 100
		}
		if numDice < 1 {
			rollText += fmt.Sprintf("%v is too few, rolling 1.\n", numDice)
			numDice = 1
		}
		sides := matches["num_sides"] // Left as string for d% rolls.
		if sides == "1" {
			rollText += "Your one-sided die rolls off into the shadows."
		} else {
			modifier := matches["modifier"]
			modifierValue, err := strconv.Atoi(matches["modifier_value"])
			if err != nil {
				modifierValue = 0 // One wasn't specified. Blame the ! modifier.
			}

			dice, total := p.RollDice(numDice, sides, modifier, modifierValue)

			if len(dice) == 1 {
				rollText += fmt.Sprintf("%q = **%d**", rollArg, total)
			} else {
				rollText += fmt.Sprintf("%q %v = **%d**", rollArg, dice, total)
			}
		}
	} else {
		rollText += fmt.Sprintf("I have no idea what to do with this: %v", rollArg)
	}

	return rollText
}

// -----------------------------------------------------------------------------
// Utility functions.
// -----------------------------------------------------------------------------

// Init - Initialize plugin.
func (p *RollyPlugin) Init() {
	// Dice rolling patterns.
	//
	// See prototype.py for more readable (?) versions of these.
	p.simplePattern = regexp.MustCompile(simpleRegex)
	p.comboPattern = regexp.MustCompile(comboRegex)
	p.rollPattern = regexp.MustCompile(rollRegex)
}

// GetCommand - Return the Command to register.
func (p *RollyPlugin) GetCommand() *model.Command {
	return &model.Command{
		Trigger:          trigger,
		Description:      "Roll one or more dice. With combos!",
		DisplayName:      pluginName,
		AutoComplete:     true,
		AutoCompleteDesc: "🎲 Roll the dice! Use `/" + trigger + " help` for usage.",
		AutoCompleteHint: "6 d10 2d4+2 and other modifiers.",
		IconURL:          iconURI,
	}
}

// GetName - What should we call the user?
func (p *RollyPlugin) GetName(userID string) (string, *model.AppError) {
	user, userErr := p.API.GetUser(userID)
	if userErr != nil {
		return "Error McErrorface", userErr
	}

	name := user.Nickname
	if len(name) < 1 {
		name = user.Username

		if len(name) < 1 {
			name = user.FirstName + " " + user.LastName
		}
	}

	return name, nil
}

// Is there already a way to do this?
func sum(values []int) int {
	total := 0
	for idx := 0; idx < len(values); idx++ {
		total += values[idx]
	}

	return total
}

// Is there already a way to do this?
func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

// FindNamedSubstrings - Return a map of named matches.
func FindNamedSubstrings(re *regexp.Regexp, candidate string) map[string]string {
	found := make(map[string]string)

	values := re.FindStringSubmatch(candidate)
	keys := re.SubexpNames()

	// Why do you start indexing keys at 1 instead of 0?
	for idx := 1; idx < len(keys); idx++ {
		found[keys[idx]] = values[idx]
	}

	return found
}

// -----------------------------------------------------------------------------
// Random number generator.
//
// Created as methods rather than just calling rand.Seed(), etc. directly
// because I'd like to make a fork of this that uses the ISARA Radiate toolkit
// RNGs to generate cryptographically secure random numbers. Massive overkill
// for this sort of application!
// -----------------------------------------------------------------------------

// RollDice - Roll {dice}d{sides}{modifier}{modifier_value}.
//
// Returns an array of rolls, and the (modified) total.
func (p *RollyPlugin) RollDice(dice int, sides string, modifier string, modifierValue int) ([]int, int) {
	var rolls []int

	// Valid dieSides are digits, or %.
	var dieSides int
	if sides == "%" {
		dieSides = 100
	} else if sides == "F" {
		dieSides = 3
	} else {
		dieSides, _ = strconv.Atoi(sides)
		if dieSides < 2 {
			dieSides = 2
		}
	}

	for idx := 0; idx < dice; idx++ {
		value := p.GetRandom(dieSides)
		if sides == "F" {
			value -= 2 // FUDGE dice product -1, 0, 1
		}
		rolls = append(rolls, value)
	}

	sort.Ints(rolls)
	total := sum(rolls)

	// Most of the supported modifiers are trivial.
	switch modifier {
	case "+":
		total += modifierValue
	case "-":
		total -= modifierValue
		if total < 1 && sides != "F" {
			total = 1 // Clamp to 1, unless FUDGE.
		}
	case "/":
		if modifierValue > 0 {
			total /= modifierValue
		}
	case "x", "*":
		total *= modifierValue
	case "<": // Ignore the lowest modifierValue rolls.
		cutoff := min(modifierValue, len(rolls)-1)

		total = sum(rolls[cutoff:])
	case ">": // Keep the best modifierValue rolls.
		if modifierValue >= len(rolls) {
			total = sum(rolls)
		} else if modifierValue < 1 {
			total = 0
		} else {
			cutoff := len(rolls) - modifierValue

			total = sum(rolls[cutoff:])
		}
	case "!": // Exploding dice!
		explode := 0
		for idx := len(rolls) - 1; idx >= 0; idx-- {
			if (sides == "F" && rolls[idx] == 1) || rolls[idx] == dieSides {
				explode++
			}
		}

		for idx := 0; idx < explode; idx++ {
			boom := p.GetRandom(dieSides)
			if sides == "F" { // FUDGE is a special case.
				boom -= 2
				if boom == 1 {
					explode++
				}
			} else { // Normal case.
				if boom == dieSides {
					explode++
				}
			}
			rolls = append(rolls, boom)
		}

		total = sum(rolls)
	}

	return rolls, total
}

// SeedRng - Seed the random number generator.
func (p *RollyPlugin) SeedRng() {
	rand.Seed(time.Now().UnixNano())
}

// GetRandom - Gets a random number from [1, n].
func (p *RollyPlugin) GetRandom(n int) int {
	return rand.Intn(n) + 1
}
